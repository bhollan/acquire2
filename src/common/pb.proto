syntax = "proto3";

message PB {
  enum ErrorCode {
    NOT_USING_LATEST_VERSION = 0;
    INTERNAL_SERVER_ERROR = 1;
    INVALID_MESSAGE_FORMAT = 2;
    INVALID_USERNAME = 3;
    MISSING_PASSWORD = 4;
    PROVIDED_PASSWORD = 5;
    INCORRECT_PASSWORD = 6;
    INVALID_MESSAGE = 7;
    COULD_NOT_CONNECT = 8;
  }

  enum GameBoardType {
    LUXOR = 0;
    TOWER = 1;
    AMERICAN = 2;
    FESTIVAL = 3;
    WORLDWIDE = 4;
    CONTINENTAL = 5;
    IMPERIAL = 6;
    NOTHING = 7;
    NOTHING_YET = 8;
    CANT_PLAY_EVER = 9;
    I_HAVE_THIS = 10;
    WILL_PUT_LONELY_TILE_DOWN = 11;
    HAVE_NEIGHBORING_TILE_TOO = 12;
    WILL_FORM_NEW_CHAIN = 13;
    WILL_MERGE_CHAINS = 14;
    CANT_PLAY_NOW = 15;
    MAX = 16;
  }

  enum GameMode {
    UNKNOWN = 0; // not used, here because the first enum value must be zero in proto3
    SINGLES_1 = 1;
    SINGLES_2 = 2;
    SINGLES_3 = 3;
    SINGLES_4 = 4;
    SINGLES_5 = 5;
    SINGLES_6 = 6;
    TEAMS_2_VS_2 = 7;
    TEAMS_2_VS_2_VS_2 = 8;
    TEAMS_3_VS_3 = 9;
  }

  enum PlayerArrangementMode {
    VERSION_1 = 0;
    RANDOM_ORDER = 1;
    EXACT_ORDER = 2;
    SPECIFY_TEAMS = 3;
  }

  message GameSetup {
    message Position {
      int32 user_id = 1;
      bool is_host = 2;
      bool approves_of_game_setup = 3;
    }

    int32 game_id = 1;
    int32 game_display_number = 2;
    GameMode game_mode = 4;
    PlayerArrangementMode player_arrangement_mode = 5;
    repeated Position positions = 6;
    repeated int32 game_board = 7;
  }

  message GameState {
    message RevealedTileRackTile {
      int32 tile = 1;
      int32 player_id_belongs_to = 2;
    }

    GameAction game_action = 1;
    int64 timestamp = 2 [ jstype = JS_NUMBER ];
    repeated RevealedTileRackTile revealed_tile_rack_tiles = 3;
    repeated int32 revealed_tile_bag_tiles = 4;
    int32 player_id_with_playable_tile_plus_one = 5; // 1 more than actual player_id
  }

  message GameSetupAction {
    message JoinGame {
    }

    message UnjoinGame {
    }

    message ApproveOfGameSetup {
    }

    message ChangeGameMode {
      GameMode game_mode = 1;
    }

    message ChangePlayerArrangementMode {
      PlayerArrangementMode player_arrangement_mode = 1;
    }

    message SwapPositions {
      int32 position1 = 1;
      int32 position2 = 2;
    }

    message KickUser {
      int32 user_id = 1;
    }

    JoinGame join_game = 1;
    UnjoinGame unjoin_game = 2;
    ApproveOfGameSetup approve_of_game_setup = 3;
    ChangeGameMode change_game_mode = 4;
    ChangePlayerArrangementMode change_player_arrangement_mode = 5;
    SwapPositions swap_positions = 6;
    KickUser kick_user = 7;
  }

  message GameSetupChange {
    message UserAdded {
      int32 user_id = 1;
    }

    message UserRemoved {
      int32 user_id = 1;
    }

    message UserApprovedOfGameSetup {
      int32 user_id = 1;
      bool approved_by_everybody = 2;
      repeated int32 final_user_ids = 3;
    }

    message GameModeChanged {
      GameMode game_mode = 1;
    }

    message PlayerArrangementModeChanged {
      PlayerArrangementMode player_arrangement_mode = 1;
    }

    message PositionsSwapped {
      int32 position1 = 1;
      int32 position2 = 2;
    }

    message UserKicked {
      int32 user_id = 1;
    }

    UserAdded user_added = 1;
    UserRemoved user_removed = 2;
    UserApprovedOfGameSetup user_approved_of_game_setup = 3;
    GameModeChanged game_mode_changed = 4;
    PlayerArrangementModeChanged player_arrangement_mode_changed = 5;
    PositionsSwapped positions_swapped = 6;
    UserKicked user_kicked = 7;
  }

  message GameAction {
    message StartGame {
    }

    message PlayTile {
      int32 tile = 1;
    }

    message SelectNewChain {
      GameBoardType chain = 1;
    }

    message SelectMergerSurvivor {
      GameBoardType chain = 1;
    }

    message SelectChainToDisposeOfNext {
      GameBoardType chain = 1;
    }

    message DisposeOfShares {
      int32 trade_amount = 1;
      int32 sell_amount = 2;
    }

    message PurchaseShares {
      repeated GameBoardType chains = 1;
      bool end_game = 2;
    }

    message GameOver {
    }

    StartGame start_game = 1;
    PlayTile play_tile = 2;
    SelectNewChain select_new_chain = 3;
    SelectMergerSurvivor select_merger_survivor = 4;
    SelectChainToDisposeOfNext select_chain_to_dispose_of_next = 5;
    DisposeOfShares dispose_of_shares = 6;
    PurchaseShares purchase_shares = 7;
    GameOver game_over = 8;
  }

  message MessageToServer {
  }

  message MessageToClient {
  }

  message MessagesToClient {
    repeated MessageToClient messages_to_client = 1;
  }

  message GameReview {
    GameMode game_mode = 1;
    PlayerArrangementMode player_arrangement_mode = 2;
    repeated int32 user_ids = 3;
    repeated string usernames = 4;
    int32 host_user_id = 5;
    repeated int32 tile_bag = 6;
    repeated GameAction game_actions = 7;
    int64 begin_timestamp = 8 [ jstype = JS_NUMBER ];
    repeated int64 game_action_timestamp_offsets = 9 [ jstype = JS_NUMBER ];
    int64 end_timestamp = 10 [ jstype = JS_NUMBER ];
  }
}
