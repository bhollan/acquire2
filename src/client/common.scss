// in svelte's global.css
// BEGIN

.luxor {
  background-color: #ff0000;
  color: #ffffff;
}

.tower {
  background-color: #ffff00;
}

.american {
  background-color: #0000ff;
  color: #ffffff;
}

.festival {
  background-color: #00ff00;
}

.worldwide {
  background-color: #804000;
  color: #ffffff;
}

.continental {
  background-color: #00ffff;
}

.imperial {
  background-color: #ff00ff;
  color: #ffffff;
}

.nothing {
  background-color: #ffffc0;
}

.nothingYet {
  background-color: #000000;
  color: #ffffff;
}

.cantPlayEver {
  background-color: #808080;
  color: #ffffff;
}

.iHaveThis {
  background-color: #ffe0a0;
}

.willPutLonelyTileDown {
  background-color: #c0c0c0;
}

.haveNeighboringTileToo {
  background-color: #c0ffc0;
}

.willFormNewChain {
  background-color: #ffffff;
}

.willMergeChains {
  background: url(./merge.png);
}

.cantPlayNow {
  background-color: #606060;
  color: #ffffff;
}

// END

// in svelte's Username component
.username {
  background-color: #d3d3d3;
}

.hotelButton {
  font-weight: bold;
  padding: 1px;

  &:focus {
    box-shadow: 0px 0px 2px 0px #323232;
  }
}

.invisible {
  visibility: hidden;
}

// in svelte's global.css
// BEGIN

.team1 {
  background-color: #fff7e7;
}

.team2 {
  background-color: #eee7ff;
}

.team3 {
  background-color: #ebffe7;
}

// END
