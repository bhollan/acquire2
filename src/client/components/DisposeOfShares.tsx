import React from 'react';
import type { PB_GameBoardType } from '../../common/pb';
import { gameBoardTypeToCSSClassName } from '../helpers';
import * as style from './DisposeOfShares.scss';

export interface DisposeOfSharesProps {
  defunctChain: PB_GameBoardType;
  controllingChain: PB_GameBoardType;
  sharesOwnedInDefunctChain: number;
  sharesAvailableInControllingChain: number;
  buttonSize: number;
  keyboardShortcutsEnabled: boolean;
  onSharesDisposed: (traded: number, sold: number) => void;
}

interface DisposeOfSharesState {
  props: DisposeOfSharesProps;
  keep: number;
  trade: number;
  tradeMax: number;
  sell: number;
  sellMax: number;
}

const keyboardShortcutToButtonIndex = new Map([
  ['1', 0],
  ['k', 0],
  ['2', 1],
  ['t', 1],
  ['3', 2],
  ['T', 2],
  ['4', 3],
  ['5', 4],
  ['s', 4],
  ['6', 5],
  ['S', 5],
  ['7', 6],
  ['0', 7],
  ['8', 7],
  ['o', 7],
]);

export class DisposeOfShares extends React.Component<DisposeOfSharesProps, DisposeOfSharesState> {
  buttonRefs: React.RefObject<HTMLInputElement>[];

  constructor(props: DisposeOfSharesProps) {
    super(props);

    this.buttonRefs = new Array(8);
    for (let i = 0; i < 8; i++) {
      this.buttonRefs[i] = React.createRef();
    }

    this.state = DisposeOfShares._getDerivedStateFromProps(props);
  }

  static _getDerivedStateFromProps(props: DisposeOfSharesProps): DisposeOfSharesState {
    return {
      props,
      keep: props.sharesOwnedInDefunctChain,
      trade: 0,
      tradeMax: Math.min(Math.floor(props.sharesOwnedInDefunctChain / 2) * 2, props.sharesAvailableInControllingChain * 2),
      sell: 0,
      sellMax: props.sharesOwnedInDefunctChain,
    };
  }

  static getDerivedStateFromProps(nextProps: DisposeOfSharesProps, prevState: DisposeOfSharesState) {
    if (
      nextProps.defunctChain !== prevState.props.defunctChain ||
      nextProps.controllingChain !== prevState.props.controllingChain ||
      nextProps.sharesOwnedInDefunctChain !== prevState.props.sharesOwnedInDefunctChain ||
      nextProps.sharesAvailableInControllingChain !== prevState.props.sharesAvailableInControllingChain
    ) {
      return DisposeOfShares._getDerivedStateFromProps(nextProps);
    } else {
      return null;
    }
  }

  shouldComponentUpdate(nextProps: DisposeOfSharesProps, nextState: DisposeOfSharesState) {
    // everything except props.keyboardShortcutsEnabled and state.props
    return (
      nextProps.defunctChain !== this.props.defunctChain ||
      nextProps.controllingChain !== this.props.controllingChain ||
      nextProps.sharesOwnedInDefunctChain !== this.props.sharesOwnedInDefunctChain ||
      nextProps.sharesAvailableInControllingChain !== this.props.sharesAvailableInControllingChain ||
      nextProps.buttonSize !== this.props.buttonSize ||
      nextProps.onSharesDisposed !== this.props.onSharesDisposed ||
      nextState.keep !== this.state.keep ||
      nextState.trade !== this.state.trade ||
      nextState.tradeMax !== this.state.tradeMax ||
      nextState.sell !== this.state.sell ||
      nextState.sellMax !== this.state.sellMax
    );
  }

  componentDidMount() {
    // do this check so tests pass
    if (typeof window !== 'undefined') {
      window.addEventListener('keydown', this.onWindowKeydown);
    }
  }

  componentWillUnmount() {
    // do this check so tests pass
    if (typeof window !== 'undefined') {
      window.removeEventListener('keydown', this.onWindowKeydown);
    }
  }

  onWindowKeydown = (event: KeyboardEvent) => {
    if (!this.props.keyboardShortcutsEnabled) {
      return;
    }

    const keyName = event.key;

    if (keyboardShortcutToButtonIndex.has(keyName)) {
      const buttonIndex = keyboardShortcutToButtonIndex.get(keyName)!;
      const button = this.buttonRefs[buttonIndex].current;
      if (button !== null) {
        button.focus();
        if (buttonIndex < 7) {
          button.click();
        }
      }
    }
  };

  render() {
    const { defunctChain, controllingChain, sharesOwnedInDefunctChain, buttonSize } = this.props;
    const { keep, trade, tradeMax, sell, sellMax } = this.state;

    const keepAllDisabled = keep === sharesOwnedInDefunctChain;
    const tradeIncrementAndMaxDisabled = trade === tradeMax;
    const tradeDecrementDisabled = trade === 0;
    const sellIncrementAndMaxDisabled = sell === sellMax;
    const sellDecrementDisabled = sell === 0;

    return (
      <div className={style.root} style={{ fontSize: Math.floor(buttonSize * 0.4) }}>
        <fieldset className={gameBoardTypeToCSSClassName.get(defunctChain)}>
          <legend className={gameBoardTypeToCSSClassName.get(defunctChain)}>Keep</legend>
          <span>{keep}</span>
          <input type={'button'} value={'All'} disabled={keepAllDisabled} ref={this.buttonRefs[0]} onClick={keepAllDisabled ? undefined : this.handleKeepAll} />
        </fieldset>
        <fieldset className={gameBoardTypeToCSSClassName.get(controllingChain)}>
          <legend className={gameBoardTypeToCSSClassName.get(controllingChain)}>Trade</legend>
          <span>{trade}</span>
          <input
            type={'button'}
            value={'▲'}
            disabled={tradeIncrementAndMaxDisabled}
            ref={this.buttonRefs[1]}
            onClick={tradeIncrementAndMaxDisabled ? undefined : this.handleTradeIncrement}
          />
          <input
            type={'button'}
            value={'▼'}
            disabled={tradeDecrementDisabled}
            ref={this.buttonRefs[2]}
            onClick={tradeDecrementDisabled ? undefined : this.handleTradeDecrement}
          />
          <input
            type={'button'}
            value={'Max'}
            disabled={tradeIncrementAndMaxDisabled}
            ref={this.buttonRefs[3]}
            onClick={tradeIncrementAndMaxDisabled ? undefined : this.handleTradeMax}
          />
        </fieldset>
        <fieldset>
          <legend>Sell</legend>
          <span>{sell}</span>
          <input
            type={'button'}
            value={'▲'}
            disabled={sellIncrementAndMaxDisabled}
            ref={this.buttonRefs[4]}
            onClick={sellIncrementAndMaxDisabled ? undefined : this.handleSellIncrement}
          />
          <input
            type={'button'}
            value={'▼'}
            disabled={sellDecrementDisabled}
            ref={this.buttonRefs[5]}
            onClick={sellDecrementDisabled ? undefined : this.handleSellDecrement}
          />
          <input
            type={'button'}
            value={'Max'}
            disabled={sellIncrementAndMaxDisabled}
            ref={this.buttonRefs[6]}
            onClick={sellIncrementAndMaxDisabled ? undefined : this.handleSellMax}
          />
        </fieldset>
        <input type={'button'} value={'OK'} ref={this.buttonRefs[7]} onClick={this.handleOK} />
      </div>
    );
  }

  handleKeepAll = () => this.determineNewState(-this.state.trade, -this.state.sell);
  handleTradeIncrement = () => this.determineNewState(2, 0);
  handleTradeDecrement = () => this.determineNewState(-2, 0);
  handleTradeMax = () => this.determineNewState(this.state.tradeMax - this.state.trade, 0);
  handleSellIncrement = () => this.determineNewState(0, 1);
  handleSellDecrement = () => this.determineNewState(0, -1);
  handleSellMax = () => this.determineNewState(0, this.state.sellMax - this.state.sell);
  handleOK = () => this.props.onSharesDisposed(this.state.trade, this.state.sell);

  determineNewState(tradeChange: number, sellChange: number) {
    const trade = this.state.trade + tradeChange;
    const sell = this.state.sell + sellChange;

    const keep = this.props.sharesOwnedInDefunctChain - trade - sell;
    const tradeMax = Math.min(trade + Math.floor(keep / 2) * 2, this.props.sharesAvailableInControllingChain * 2);
    const sellMax = sell + keep;

    this.setState({ keep, trade, tradeMax, sell, sellMax });
  }
}
