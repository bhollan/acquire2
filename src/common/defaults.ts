import type { GameState } from './game';
import { PB_GameBoardType } from './pb';

export function defaultGameStateHistory(): GameState[] {
	return [];
}

export function defaultTileRacks(): (number | null)[][] {
	return [];
}

export function defaultTileRack(): (number | null)[] {
	return [null, null, null, null, null, null];
}

export function defaultTileRackTypesList(): (PB_GameBoardType | null)[][] {
	return [];
}

export function defaultTileRackTypes(): (PB_GameBoardType | null)[] {
	return [null, null, null, null, null, null];
}

const defaultGameBoardRow = [
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
	PB_GameBoardType.NOTHING,
];
export function defaultGameBoard(): PB_GameBoardType[][] {
	return [
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
		[...defaultGameBoardRow],
	];
}

export function defaultScoreBoard(): number[][] {
	return [];
}

export function defaultScoreBoardRow(): number[] {
	return [0, 0, 0, 0, 0, 0, 0, 60, 60];
}

export function defaultScoreBoardAvailable(): number[] {
	return [25, 25, 25, 25, 25, 25, 25];
}

export function defaultScoreBoardChainSize(): number[] {
	return [0, 0, 0, 0, 0, 0, 0];
}

export function defaultScoreBoardPrice(): number[] {
	return [0, 0, 0, 0, 0, 0, 0];
}

export function defaultSafeChains(): boolean[] {
	return [false, false, false, false, false, false, false];
}
