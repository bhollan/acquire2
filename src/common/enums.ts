export enum GameActionEnum {
	StartGame,
	PlayTile,
	SelectNewChain,
	SelectMergerSurvivor,
	SelectChainToDisposeOfNext,
	DisposeOfShares,
	PurchaseShares,
	GameOver,
}

export enum GameHistoryMessageEnum {
	TurnBegan,
	DrewPositionTile,
	StartedGame,
	DrewTile,
	HasNoPlayableTile,
	PlayedTile,
	FormedChain,
	MergedChains,
	SelectedMergerSurvivor,
	SelectedChainToDisposeOfNext,
	ReceivedBonus,
	DisposedOfShares,
	CouldNotAffordAnyShares,
	PurchasedShares,
	DrewLastTile,
	ReplacedDeadTile,
	EndedGame,
	NoTilesPlayedForEntireRound,
	AllTilesPlayed,
}

export enum ScoreBoardIndexEnum {
	Luxor,
	Tower,
	American,
	Festival,
	Worldwide,
	Continental,
	Imperial,
	Cash,
	Net,
	Max,
}

export enum TileEnum {
	Unknown = 108,
}
