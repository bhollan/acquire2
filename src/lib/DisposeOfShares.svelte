<svelte:options immutable />

<script lang="ts" context="module">
  const keyboardShortcutToButtonIndex = new Map([
    ['1', 0],
    ['k', 0],
    ['2', 1],
    ['t', 1],
    ['3', 2],
    ['T', 2],
    ['4', 3],
    ['5', 4],
    ['s', 4],
    ['6', 5],
    ['S', 5],
    ['7', 6],
    ['0', 7],
    ['8', 7],
    ['o', 7],
  ]);
</script>

<script lang="ts">
  import type { PB_GameBoardType } from '../common/pb';
  import { gameBoardTypeToCSSClassName } from './helpers';

  export let defunctChain: PB_GameBoardType;
  export let controllingChain: PB_GameBoardType;
  export let sharesOwnedInDefunctChain: number;
  export let sharesAvailableInControllingChain: number;
  export let buttonSize: number;
  export let keyboardShortcutsEnabled: boolean;
  export let onSharesDisposed: (traded: number, sold: number) => void;

  let keep = 0;
  let trade = 0;
  let tradeMax = 0;
  let sell = 0;
  let sellMax = 0;
  updateValues(0, 0);

  const inputs: (HTMLInputElement | undefined)[] = new Array(8);
  inputs.fill(undefined);

  function updateValues(tradeChange: number, sellChange: number) {
    trade = trade + tradeChange;
    sell = sell + sellChange;

    keep = sharesOwnedInDefunctChain - trade - sell;
    tradeMax = Math.min(trade + Math.floor(keep / 2) * 2, sharesAvailableInControllingChain * 2);
    sellMax = sell + keep;
  }

  function handleKeydown(event: KeyboardEvent) {
    if (keyboardShortcutsEnabled) {
      const buttonIndex = keyboardShortcutToButtonIndex.get(event.key);

      if (buttonIndex !== undefined) {
        const button = inputs[buttonIndex];

        if (button) {
          button.focus();
          if (buttonIndex < 7) {
            button.click();
          }
        }
      }
    }
  }
</script>

<svelte:window on:keydown={handleKeydown} />

<div style="font-size: {Math.floor(buttonSize * 0.4)}px;">
  <fieldset class={gameBoardTypeToCSSClassName.get(defunctChain)}>
    <legend class={gameBoardTypeToCSSClassName.get(defunctChain)}>Keep</legend>
    <span>{keep}</span>
    <input type="button" value="All" disabled={keep === sharesOwnedInDefunctChain} bind:this={inputs[0]} on:click={() => updateValues(-trade, -sell)} />
  </fieldset>
  <fieldset class={gameBoardTypeToCSSClassName.get(controllingChain)}>
    <legend class={gameBoardTypeToCSSClassName.get(controllingChain)}>Trade</legend>
    <span>{trade}</span>
    <input type="button" value="▲" disabled={trade === tradeMax} bind:this={inputs[1]} on:click={() => updateValues(2, 0)} />
    <input type="button" value="▼" disabled={trade === 0} bind:this={inputs[2]} on:click={() => updateValues(-2, 0)} />
    <input type="button" value="Max" disabled={trade === tradeMax} bind:this={inputs[3]} on:click={() => updateValues(tradeMax - trade, 0)} />
  </fieldset>
  <fieldset>
    <legend>Sell</legend>
    <span>{sell}</span>
    <input type="button" value="▲" disabled={sell === sellMax} bind:this={inputs[4]} on:click={() => updateValues(0, 1)} />
    <input type="button" value="▼" disabled={sell === 0} bind:this={inputs[5]} on:click={() => updateValues(0, -1)} />
    <input type="button" value="Max" disabled={sell === sellMax} bind:this={inputs[6]} on:click={() => updateValues(0, sellMax - sell)} />
  </fieldset>
  <input type="button" value="OK" bind:this={inputs[7]} on:click={() => onSharesDisposed(trade, sell)} />
</div>

<style>
  fieldset {
    display: inline-block;
    margin: 0;
    padding: 0 3px 3px;
  }

  legend {
    padding: 0 5px;
  }

  span {
    display: inline-block;
    text-align: right;
    width: 1.5em;
  }

  input {
    color: #000000;
  }

  input:disabled {
    color: #808080;
  }
</style>
