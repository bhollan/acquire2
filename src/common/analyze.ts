import fs from 'fs';
import path from 'path';
import { GameHistoryMessageEnum } from './enums';
import { gameFromProtocolBuffer } from './gameSerialization';
import { PB_GameReview } from './pb';
import { getGameHistoryMessageString } from './runGameTestFile';

function main(gameFilePath: string) {
  const gameReviewFileContents = fs.readFileSync(gameFilePath);
  const game = gameFromProtocolBuffer(PB_GameReview.fromBinary(gameReviewFileContents));

  for (let i = 0; i < game.gameStateHistory.length; i++) {
    console.log(i);

    const gameState = game.gameStateHistory[i];
    for (const gameHistoryMessage of gameState.gameHistoryMessages) {
      console.log(getGameHistoryMessageString(gameHistoryMessage));
    }
  }

  console.log();
  console.log();
  console.log();

  for (let i = 0; i < game.gameStateHistory.length; i++) {
    const gameState = game.gameStateHistory[i];
    const mergedChains = gameState.gameHistoryMessages.filter((m) => m.gameHistoryMessage === GameHistoryMessageEnum.ReceivedBonus).length > 0;

    if (mergedChains) {
      console.log(i);

      for (const gameHistoryMessage of gameState.gameHistoryMessages) {
        console.log(getGameHistoryMessageString(gameHistoryMessage));
      }
    }
  }

  updateReviewGamePBBinary(gameReviewFileContents);
}

function updateReviewGamePBBinary(gameReviewFileContents: Buffer) {
  const pathToFile = path.join(__dirname, '../../src/client/reviewGamePBBinary.ts');

  const currentContents = fs.readFileSync(pathToFile).toString();
  const desiredContents = `export const reviewGamePBBinary = new Uint8Array(${JSON.stringify([...gameReviewFileContents])});\n`;

  if (desiredContents !== currentContents) {
    fs.writeFileSync(pathToFile, desiredContents);
  }
}

main(process.argv[2]);
