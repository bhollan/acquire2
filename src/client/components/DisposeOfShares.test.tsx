import { shallow } from 'enzyme';
import React from 'react';
import * as renderer from 'react-test-renderer';
import { PB_GameBoardType } from '../../common/pb';
import { configureEnzyme } from '../configureEnzyme';
import { DisposeOfShares } from './DisposeOfShares';

configureEnzyme();

enum inputIndex {
  keepAll,
  tradeIncrement,
  tradeDecrement,
  tradeMax,
  sellIncrement,
  sellDecrement,
  sellMax,
  ok,
}

function getNewComponent(
  sharesOwnedInDefunctChain: number,
  sharesAvailableInControllingChain: number,
  onSharesDisposed: (traded: number, sold: number) => void,
) {
  return shallow(
    <DisposeOfShares
      defunctChain={PB_GameBoardType.IMPERIAL}
      controllingChain={PB_GameBoardType.TOWER}
      sharesOwnedInDefunctChain={sharesOwnedInDefunctChain}
      sharesAvailableInControllingChain={sharesAvailableInControllingChain}
      buttonSize={40}
      keyboardShortcutsEnabled={false}
      onSharesDisposed={onSharesDisposed}
    />,
  );
}

function clickInput(component: any, index: inputIndex) {
  component.find('input').at(index).simulate('click');
}

function expectValues(component: any, keepValue: number, tradeValue: number, sellValue: number) {
  const fieldsets = component.find('fieldset');
  expect(fieldsets.at(0).text()).toBe(`Keep${keepValue}`);
  expect(fieldsets.at(1).text()).toBe(`Trade${tradeValue}`);
  expect(fieldsets.at(2).text()).toBe(`Sell${sellValue}`);
}

test('renders correctly', () => {
  const onSharesDisposed = () => {
    // do nothing
  };

  const component = renderer.create(
    <DisposeOfShares
      defunctChain={PB_GameBoardType.IMPERIAL}
      controllingChain={PB_GameBoardType.TOWER}
      sharesOwnedInDefunctChain={7}
      sharesAvailableInControllingChain={2}
      buttonSize={40}
      keyboardShortcutsEnabled={false}
      onSharesDisposed={onSharesDisposed}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('works correctly', () => {
  const onSharesDisposed = jest.fn();
  const component = getNewComponent(7, 2, onSharesDisposed);

  expectValues(component, 7, 0, 0);
  clickInput(component, inputIndex.tradeIncrement);
  expectValues(component, 5, 2, 0);
  clickInput(component, inputIndex.tradeIncrement);
  expectValues(component, 3, 4, 0);
  clickInput(component, inputIndex.tradeIncrement);
  expectValues(component, 3, 4, 0);
  clickInput(component, inputIndex.ok);
  expect(onSharesDisposed.mock.calls.length).toBe(1);
  expect(onSharesDisposed.mock.calls[0]).toEqual([4, 0]);

  clickInput(component, inputIndex.keepAll);
  expectValues(component, 7, 0, 0);
  clickInput(component, inputIndex.tradeMax);
  expectValues(component, 3, 4, 0);
  clickInput(component, inputIndex.tradeIncrement);
  expectValues(component, 3, 4, 0);
  clickInput(component, inputIndex.sellMax);
  expectValues(component, 0, 4, 3);
  clickInput(component, inputIndex.ok);
  expect(onSharesDisposed.mock.calls.length).toBe(2);
  expect(onSharesDisposed.mock.calls[1]).toEqual([4, 3]);

  clickInput(component, inputIndex.keepAll);
  expectValues(component, 7, 0, 0);
  clickInput(component, inputIndex.sellIncrement);
  clickInput(component, inputIndex.sellIncrement);
  clickInput(component, inputIndex.sellIncrement);
  clickInput(component, inputIndex.sellIncrement);
  expectValues(component, 3, 0, 4);
  clickInput(component, inputIndex.tradeMax);
  expectValues(component, 1, 2, 4);
  clickInput(component, inputIndex.tradeIncrement);
  expectValues(component, 1, 2, 4);
  clickInput(component, inputIndex.sellMax);
  expectValues(component, 0, 2, 5);
  clickInput(component, inputIndex.ok);
  expect(onSharesDisposed.mock.calls.length).toBe(3);
  expect(onSharesDisposed.mock.calls[2]).toEqual([2, 5]);

  clickInput(component, inputIndex.keepAll);
  expectValues(component, 7, 0, 0);
  clickInput(component, inputIndex.sellMax);
  clickInput(component, inputIndex.sellDecrement);
  expectValues(component, 1, 0, 6);
  clickInput(component, inputIndex.tradeMax);
  expectValues(component, 1, 0, 6);
  clickInput(component, inputIndex.ok);
  expect(onSharesDisposed.mock.calls.length).toBe(4);
  expect(onSharesDisposed.mock.calls[3]).toEqual([0, 6]);

  clickInput(component, inputIndex.keepAll);
  expectValues(component, 7, 0, 0);
  clickInput(component, inputIndex.ok);
  expect(onSharesDisposed.mock.calls.length).toBe(5);
  expect(onSharesDisposed.mock.calls[4]).toEqual([0, 0]);
});
