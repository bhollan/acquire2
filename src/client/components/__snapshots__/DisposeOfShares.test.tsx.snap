// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="root"
  style={
    {
      "fontSize": 16,
    }
  }
>
  <fieldset
    className="imperial"
  >
    <legend
      className="imperial"
    >
      Keep
    </legend>
    <span>
      7
    </span>
    <input
      disabled={true}
      type="button"
      value="All"
    />
  </fieldset>
  <fieldset
    className="tower"
  >
    <legend
      className="tower"
    >
      Trade
    </legend>
    <span>
      0
    </span>
    <input
      disabled={false}
      onClick={[Function]}
      type="button"
      value="▲"
    />
    <input
      disabled={true}
      type="button"
      value="▼"
    />
    <input
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Max"
    />
  </fieldset>
  <fieldset>
    <legend>
      Sell
    </legend>
    <span>
      0
    </span>
    <input
      disabled={false}
      onClick={[Function]}
      type="button"
      value="▲"
    />
    <input
      disabled={true}
      type="button"
      value="▼"
    />
    <input
      disabled={false}
      onClick={[Function]}
      type="button"
      value="Max"
    />
  </fieldset>
  <input
    onClick={[Function]}
    type="button"
    value="OK"
  />
</div>
`;
